File: crates/nxvim-lua/src/prelude/promise.lua
-- nxvim Lua prelude — nx.promise, a Promises/A+ surface shaped like the browser's.
--
-- `nx` async is PROMISE-ONLY (ADR 0002): a one-shot async API returns a promise
-- (nx.run, nx.fs, …) and streaming is an async-iterator over them (nx.run_stream +
-- nx.await_each). `nx.promise` is the foundation: the exact object model the
-- browser exposes — `nx.promise.new(executor)`, `:next`/`:catch`/`:finally`, and
-- the `nx.promise.all/all_settled/race/any/resolve/reject/try` combinators — plus
-- `nx.async`/`nx.await` coroutine sugar so a chain of awaits reads like
-- straight-line code. (`nx.promise.try` folds a function that may throw
-- SYNCHRONOUSLY or reject ASYNCHRONOUSLY into one chain — see its definition.)
--
-- This is pure Lua layered on the existing async runtime: a promise's reactions
-- run as MICROTASKS via `nx.schedule` (prelude/runtime.lua) — deferred to the end
-- of the current convergence, never inline, exactly like the browser runs `.then`
-- callbacks off the current stack. So `:next` is *always* asynchronous even for an
-- already-settled promise, matching Promises/A+ §2.2.4.
--
-- It is an nxvim-native surface (no `vim.*` twin — neovim core has no Promise), so
-- it lives entirely on `nx`. As the async FOUNDATION every later surface builds on
-- (process / picker / complete / nx.ui), it loads early — right after the runtime
-- services it needs: `nx.schedule` (the microtask primitive the reactions run on)
-- and `nx.timer` (the wall-clock deferral `nx.promise.delay` uses), both installed
-- in prelude/runtime.lua just above.

local vim = vim
nx = nx or {}

-- Run `fn` as a microtask — off the current stack, at convergence, same tick.
-- The single scheduling primitive the whole module is built on.
local function microtask(fn)
  nx.schedule(fn)
end

-- A value is "callable" if it's a function or a table with a __call metamethod —
-- the test Promises/A+ uses before treating an onFulfilled/onRejected as a handler.
local function is_callable(v)
  if type(v) == "function" then
    return true
  end
  if type(v) == "table" then
    local mt = getmetatable(v)
    return mt ~= nil and type(mt.__call) == "function"
  end
  return false
end

-- ----- the Promise object ----------------------------------------------------

local Promise = {}
Promise.__index = Promise

local PENDING, FULFILLED, REJECTED = "pending", "fulfilled", "rejected"

-- True for any of our own promises (the cheap, exact thenable check).
local function is_promise(v)
  return type(v) == "table" and getmetatable(v) == Promise
end

local function new_pending()
  return setmetatable({
    _state = PENDING,
    _value = nil,
    _queue = nil, -- list of reaction thunks while pending; nil once settled
    _handled = false, -- did anything subscribe (for unhandled-rejection reporting)
  }, Promise)
end

-- Forward declaration: the Promises/A+ resolution procedure (§2.3).
local resolve_promise

-- A rejected promise that nothing ever handles is a silent swallowed error — the
-- one thing this project refuses to ship (CLAUDE.md: no silent failures). The
-- browser reports "Uncaught (in promise)"; we do the same. After a promise
-- rejects we schedule a check one microtask later: if STILL nothing has
-- subscribed (`:next`/`:catch`/`:finally` flips `_handled`), surface it. The
-- delay gives synchronous `p:catch(...)` right after creation time to attach.
local function report_unhandled(p)
  microtask(function()
    if p._state == REJECTED and not p._handled then
      vim.notify("nxvim: unhandled promise rejection: " .. tostring(p._value), 4)
    end
  end)
end

-- Move a pending promise to its final state and flush queued reactions as
-- microtasks. A no-op if already settled (a promise settles exactly once).
local function settle(p, state, value)
  if p._state ~= PENDING then
    return
  end
  p._state = state
  p._value = value
  local queue = p._queue
  p._queue = nil
  if queue then
    for _, run in ipairs(queue) do
      microtask(run)
    end
  end
  if state == REJECTED then
    report_unhandled(p)
  end
end

-- Register reactions on `p`. If still pending they queue; if already settled the
-- matching one runs on a fresh microtask. Subscribing marks the promise handled,
-- which is what suppresses the unhandled-rejection warning.
function Promise:_subscribe(on_fulfilled, on_rejected)
  self._handled = true
  local function run()
    if self._state == FULFILLED then
      on_fulfilled(self._value)
    else
      on_rejected(self._value)
    end
  end
  if self._state == PENDING then
    self._queue = self._queue or {}
    self._queue[#self._queue + 1] = run
  else
    microtask(run)
  end
end

-- The resolution procedure: settle `p` with `x`, adopting `x`'s state if it is a
-- promise/thenable (so `resolve(anotherPromise)` waits on it) rather than
-- fulfilling with the promise object itself.
resolve_promise = function(p, x)
  if x == p then
    -- Resolving a promise with itself would deadlock — reject per §2.3.1.
    return settle(p, REJECTED, "TypeError: promise resolved with itself (chaining cycle)")
  end
  if is_promise(x) then
    return x:_subscribe(function(v)
      resolve_promise(p, v)
    end, function(e)
      settle(p, REJECTED, e)
    end)
  end
  -- Foreign thenable: anything exposing a callable `next` (our convention). Adopt
  -- it through its own `next`, guarding the access/call the way §2.3.3 requires.
  if type(x) == "table" then
    local ok, next_fn = pcall(function()
      return x.next
    end)
    if not ok then
      return settle(p, REJECTED, next_fn)
    end
    if is_callable(next_fn) then
      local called = false
      local ok2, err = pcall(next_fn, x, function(v)
        if not called then
          called = true
          resolve_promise(p, v)
        end
      end, function(e)
        if not called then
          called = true
          settle(p, REJECTED, e)
        end
      end)
      if not ok2 and not called then
        settle(p, REJECTED, err)
      end
      return
    end
  end
  settle(p, FULFILLED, x)
end

-- :next(on_fulfilled, on_rejected) — the spine. Returns a NEW promise resolved
-- with the handler's return (adopting it if it's itself a promise), or rejected
-- if the handler throws. A missing handler passes the value/reason straight
-- through to the returned promise, which is what makes a bare `:catch` at the end
-- of a chain catch errors from anywhere earlier in it.
--
-- Named `:next` (not `:then`) because `then` is a Lua keyword — `p:then(...)`
-- won't parse. `:next` is the idiomatic Lua spelling; `:catch`/`:finally` keep
-- their browser names verbatim.
function Promise:next(on_fulfilled, on_rejected)
  local result = new_pending()
  self:_subscribe(function(value)
    if is_callable(on_fulfilled) then
      local ok, r = pcall(on_fulfilled, value)
      if ok then
        resolve_promise(result, r)
      else
        settle(result, REJECTED, r)
      end
    else
      settle(result, FULFILLED, value)
    end
  end, function(reason)
    if is_callable(on_rejected) then
      local ok, r = pcall(on_rejected, reason)
      if ok then
        resolve_promise(result, r)
      else
        settle(result, REJECTED, r)
      end
    else
      settle(result, REJECTED, reason)
    end
  end)
  return result
end

-- :catch(on_rejected) — sugar for :next(nil, on_rejected), verbatim browser.
function Promise:catch(on_rejected)
  return self:next(nil, on_rejected)
end

-- :finally(on_finally) — run `on_finally` whichever way the promise settles, then
-- pass the original value/reason through untouched (so `finally` can't swallow a
-- result or a rejection). Matches the browser's pass-through semantics.
function Promise:finally(on_finally)
  return self:next(function(value)
    if is_callable(on_finally) then
      on_finally()
    end
    return value
  end, function(reason)
    if is_callable(on_finally) then
      on_finally()
    end
    error(reason, 0) -- rethrow so the rejection propagates past finally
  end)
end

-- ----- nx.promise: constructors + combinators --------------------------------

local M = {}

-- nx.promise.new(executor): the browser constructor. `executor(resolve, reject)`
-- runs SYNCHRONOUSLY now (as in the browser); a throw inside it rejects the
-- promise. `nx.promise(executor)` is the same thing via __call sugar.
function M.new(executor)
  local p = new_pending()
  if executor ~= nil then
    if not is_callable(executor) then
      error("nx.promise.new: executor must be a function", 2)
    end
    local ok, err = pcall(executor, function(value)
      resolve_promise(p, value)
    end, function(reason)
      settle(p, REJECTED, reason)
    end)
    if not ok then
      settle(p, REJECTED, err)
    end
  end
  return p
end

-- nx.promise.resolve(value): a promise already fulfilled with `value` (or, if
-- `value` is itself a promise/thenable, the very same/adopted promise).
function M.resolve(value)
  if is_promise(value) then
    return value
  end
  local p = new_pending()
  resolve_promise(p, value)
  return p
end

-- nx.promise.reject(reason): a promise already rejected with `reason`.
function M.reject(reason)
  local p = new_pending()
  settle(p, REJECTED, reason)
  return p
end

-- nx.promise.try(fn, ...): run `fn(...)` INSIDE a promise — a synchronous throw
-- becomes a rejection, and a returned promise (or plain value) is adopted. So a
-- function that may fail either way (sync error before it returns, or async
-- rejection of what it returns) folds into ONE chain: no `pcall` + branch at the
-- call site, just `nx.promise.try(fn, ...):next(...):catch(...)`. Mirrors the
-- browser's `Promise.try`.
function M.try(fn, ...)
  local args = { ... }
  local argc = select("#", ...)
  return M.new(function(resolve)
    -- A throw in `fn` propagates out of this executor, which M.new turns into a
    -- rejection; a returned promise is adopted by `resolve` (the resolution proc).
    resolve(fn(table.unpack(args, 1, argc)))
  end)
end

-- Count a 1..n list (combinators take array-like tables of promises/values).
local function list_len(t)
  return #t
end

-- nx.promise.all(list): fulfils with the array of every value once ALL fulfil, in
-- input order; rejects as soon as ANY rejects (with that reason). An empty list
-- fulfils immediately with {}.
function M.all(list)
  return M.new(function(resolve, reject)
    local n = list_len(list)
    if n == 0 then
      return resolve({})
    end
    local results = {}
    local remaining = n
    for i = 1, n do
      M.resolve(list[i]):next(function(value)
        results[i] = value
        remaining = remaining - 1
        if remaining == 0 then
          resolve(results)
        end
      end, reject)
    end
  end)
end

-- nx.promise.all_settled(list) [alias allSettled]: fulfils once every promise
-- settles, with an array of outcome tables: { status = "fulfilled", value = v }
-- or { status = "rejected", reason = e }. Never rejects.
function M.all_settled(list)
  return M.new(function(resolve)
    local n = list_len(list)
    if n == 0 then
      return resolve({})
    end
    local results = {}
    local remaining = n
    local function record(i, outcome)
      results[i] = outcome
      remaining = remaining - 1
      if remaining == 0 then
        resolve(results)
      end
    end
    for i = 1, n do
      M.resolve(list[i]):next(function(value)
        record(i, { status = "fulfilled", value = value })
      end, function(reason)
        record(i, { status = "rejected", reason = reason })
      end)
    end
  end)
end

-- nx.promise.race(list): settles the moment the FIRST input settles, adopting its
-- fulfilment or rejection. An empty list stays pending forever (as in the
-- browser).
function M.race(list)
  return M.new(function(resolve, reject)
    for i = 1, list_len(list) do
      M.resolve(list[i]):next(resolve, reject)
    end
  end)
end

-- nx.promise.any(list): fulfils with the first value to fulfil; rejects only if
-- ALL reject, with an aggregate { errors = {...} }. An empty list rejects at once.
function M.any(list)
  return M.new(function(resolve, reject)
    local n = list_len(list)
    if n == 0 then
      return reject({ message = "All promises were rejected", errors = {} })
    end
    local errors = {}
    local remaining = n
    for i = 1, n do
      M.resolve(list[i]):next(resolve, function(reason)
        errors[i] = reason
        remaining = remaining - 1
        if remaining == 0 then
          reject({ message = "All promises were rejected", errors = errors })
        end
      end)
    end
  end)
end

-- ----- nx-native conveniences (still promise-shaped) -------------------------

-- nx.promise.delay(ms[, value]): a promise that fulfils with `value` after `ms`
-- wall-clock milliseconds, on the loop. The promise-flavoured vim.defer_fn — the
-- await-able sleep that makes retry/debounce chains read linearly.
function M.delay(ms, value)
  return M.new(function(resolve)
    nx.timer(function()
      resolve(value)
    end, ms or 0)
  end)
end

-- nx.wait_for(predicate[, opts]) -> promise: poll `predicate` BETWEEN ticks until it
-- returns a truthy value, then fulfil with that value. The await-able form of the
-- bounded "spin until a cross-tick condition holds" loop that recurs across the
-- codebase — a freshly-mounted window's id, a server-repopulated mirror, a view
-- buffer that exists next tick. It yields the tick (nx.on_next_tick) between checks, so
-- those mirrors actually refresh, instead of spinning within one convergence the way
-- a bare nx.schedule re-arm does. `predicate` is checked once immediately, then once
-- per following tick.
--
--   opts.tries     max checks before giving up (default 200 — a few seconds of ticks)
--   opts.interval  ms between checks (default: the next tick); set for slower polling
--   opts.message   the rejection message used on timeout
--
-- REJECTS (so an `nx.await` fails loud and a chain can `:catch`) if the condition
-- never holds within `tries`, or if `predicate` throws. RESOLVES with the predicate's
-- truthy value, so `:next(function(win) … end)` receives it directly. A best-effort
-- caller should `:catch` the timeout.
function M.wait_for(predicate, opts)
  opts = opts or {}
  local tries = opts.tries or 200
  local interval = opts.interval
  return M.new(function(resolve, reject)
    local n = 0
    local function step()
      local ok, val = pcall(predicate)
      if not ok then
        return reject({ message = "nx.wait_for: predicate errored: " .. tostring(val) })
      end
      if val then
        return resolve(val)
      end
      n = n + 1
      if n >= tries then
        return reject({
          message = opts.message or ("nx.wait_for: condition not met after " .. tries .. " ticks"),
        })
      end
      if interval then
        nx.timer(step, interval)
      else
        nx.on_next_tick(step)
      end
    end
    step() -- check immediately (tick 0), then poll the next ticks
  end)
end
nx.wait_for = M.wait_for

-- nx.promise.wrap(fn): lift a single-callback async function into a
-- promise-returning one. The wrapped function appends a resolver as the LAST
-- argument and resolves with whatever that callback receives (its single arg, or
-- all of them as a table when there's more than one) — the shape nxvim's own
-- callback APIs use (nx.ui.select's on_choice, an on_exit, …). Use this to turn
-- "pass me a callback" surfaces into await-ables; reach for nx.promise.new
-- directly when the callback uses a node-style (err, value) convention.
function M.wrap(fn)
  return function(...)
    local args = { ... }
    local argc = select("#", ...)
    return M.new(function(resolve)
      args[argc + 1] = function(...)
        local nres = select("#", ...)
        if nres <= 1 then
          resolve((...))
        else
          resolve({ ... })
        end
      end
      fn(table.unpack(args, 1, argc + 1))
    end)
  end
end

setmetatable(M, {
  __call = function(_, executor)
    return M.new(executor)
  end,
})

-- Browser-name aliases so muscle memory works (the canonical names are snake_case
-- to match the rest of nx.*).
M.allSettled = M.all_settled

nx.promise = M

-- ----- nx.async / nx.await: coroutine sugar over the same promises -----------
--
-- `nx.async(fn)` returns a function that, when called, runs `fn` as a coroutine
-- and returns a promise for its result. Inside, `nx.await(p)` suspends until `p`
-- settles and evaluates to its value (or re-raises its rejection as a Lua error),
-- so a sequence of awaits reads top-to-bottom with no nesting:
--
--     local load = nx.async(function(path)
--       local stat = nx.await(fs.stat(path))
--       local data = nx.await(fs.read(path))
--       return parse(data, stat)
--     end)
--     load("init.lua"):next(use):catch(report)
--
-- A rejected await raises inside the coroutine, so you can handle it either way:
-- wrap the await in `pcall` to catch it locally (PUC 5.2+ yields across a `pcall`
-- boundary, so this works on nxvim's 5.4 backend), or let it propagate to the
-- coroutine edge (the returned promise rejects, caught by `:catch` on the result)
-- or attach `:catch` to the awaited promise itself.

function nx.async(fn)
  if not is_callable(fn) then
    error("nx.async: argument must be a function", 2)
  end
  return function(...)
    local args = { ... }
    local argc = select("#", ...)
    return M.new(function(resolve, reject)
      local co = coroutine.create(fn)
      -- Drive the coroutine one resume at a time. Each yield hands us the awaited
      -- value; we resolve it to a promise and re-enter when it settles, passing
      -- (ok, value) back so nx.await can return the value or raise the reason.
      local function step(...)
        local ok, yielded = coroutine.resume(co, ...)
        if not ok then
          -- The coroutine body threw (or an awaited rejection was re-raised).
          return reject(yielded)
        end
        if coroutine.status(co) == "dead" then
          -- `fn` returned `yielded` — that's the async function's result.
          return resolve(yielded)
        end
        -- Suspended on an await: `yielded` is the awaited value/promise.
        M.resolve(yielded):next(function(value)
          step(true, value)
        end, function(reason)
          step(false, reason)
        end)
      end
      step(table.unpack(args, 1, argc))
    end)
  end
end

-- nx.await(awaitable): suspend the enclosing nx.async coroutine until `awaitable`
-- settles. Returns the fulfilment value, or raises the rejection reason as an
-- error (which, uncaught, rejects the async function's promise). Errors loudly if
-- called outside an nx.async coroutine — there is nothing to suspend.
function nx.await(awaitable)
  -- `coroutine.isyieldable()` is false on the main thread and true inside a
  -- coroutine — exactly "is there an nx.async frame to suspend?". (The 5.1
  -- spelling `coroutine.running() == nil` broke in 5.2+, where `running()`
  -- returns the main thread itself rather than nil.)
  if not coroutine.isyieldable() then
    error("nx.await must be called inside an nx.async function", 2)
  end
  local ok, value = coroutine.yield(awaitable)
  if not ok then
    error(value, 0)
  end
  return value
end

return nx

File: crates/bemtvi-server/tests/promise.rs
//! Behavior tests for `btv.promise` — the Promises/A+ surface shaped like the
//! browser's — and its `btv.async`/`btv.await` coroutine sugar. Black-box per the
//! project conventions: a real server over RPC, driven with `nvim_exec_lua`,
//! asserting on observable Lua state.
//!
//! Promise reactions run as MICROTASKS (`btv.schedule`), so they settle at the end
//! of the tick that scheduled them — within the same convergence the server runs
//! after each `nvim_exec_lua`. The pattern below therefore mirrors the
//! `vim.schedule` tests: set up a chain in one chunk writing its outcome to a
//! global, then read that global back in a second chunk (by which point every
//! microtask generation has flushed). Off-tick behaviour (`btv.promise.delay`) uses
//! the two-barrier pattern with a real sleep.

use std::time::Duration;

use bemtvi_rpc::{Incoming, Rpc};
use bemtvi_server::ServerInit;
use bemtvi_test_harness::{exec_lua, lua_bool, lua_u64, start_attached};
use tokio::sync::mpsc::UnboundedReceiver;

async fn start() -> (Rpc, UnboundedReceiver<Incoming>) {
    start_attached(ServerInit::default(), 80, 24).await
}

/// Read a `return`-style chunk as an owned `String` (`None` if not a string).
async fn lua_string(rpc: &Rpc, code: &str) -> Option<String> {
    exec_lua(rpc, code).await.as_str().map(str::to_owned)
}

// ----- core: resolve / reject / chaining -------------------------------------

#[tokio::test]
async fn resolved_value_flows_to_next() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.got = nil\n\
         btv.promise.resolve(41):next(function(v) _G.got = v + 1 end)",
    )
    .await;
    assert_eq!(lua_u64(&rpc, "return _G.got").await, Some(42));
}

#[tokio::test]
async fn rejection_is_caught() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.err = nil\n\
         btv.promise.reject('boom'):catch(function(e) _G.err = e end)",
    )
    .await;
    assert_eq!(
        lua_string(&rpc, "return _G.err").await.as_deref(),
        Some("boom")
    );
}

#[tokio::test]
async fn next_chains_transform_the_value() {
    let (rpc, _incoming) = start().await;
    // Each :next returns a new promise resolved with the handler's return; a
    // returned promise is adopted, not nested.
    exec_lua(
        &rpc,
        "_G.got = nil\n\
         btv.promise.resolve(1)\n\
           :next(function(v) return v + 1 end)\n\
           :next(function(v) return btv.promise.resolve(v * 10) end)\n\
           :next(function(v) _G.got = v end)",
    )
    .await;
    assert_eq!(lua_u64(&rpc, "return _G.got").await, Some(20));
}

#[tokio::test]
async fn a_throw_midchain_is_caught_at_the_end() {
    let (rpc, _incoming) = start().await;
    // A throw anywhere in the chain skips later :next handlers and lands in the
    // trailing :catch — the property that makes one terminal catch enough.
    exec_lua(
        &rpc,
        "_G.reached = false\n_G.err = nil\n\
         btv.promise.resolve(1)\n\
           :next(function() error('mid') end)\n\
           :next(function() _G.reached = true end)\n\
           :catch(function(e) _G.err = e end)",
    )
    .await;
    assert_eq!(lua_bool(&rpc, "return _G.reached").await, Some(false));
    assert_eq!(
        lua_bool(
            &rpc,
            "return _G.err ~= nil and string.find(_G.err, 'mid') ~= nil"
        )
        .await,
        Some(true)
    );
}

#[tokio::test]
async fn finally_runs_and_passes_the_value_through() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.ran = false\n_G.got = nil\n\
         btv.promise.resolve(7)\n\
           :finally(function() _G.ran = true end)\n\
           :next(function(v) _G.got = v end)",
    )
    .await;
    assert_eq!(lua_bool(&rpc, "return _G.ran").await, Some(true));
    assert_eq!(lua_u64(&rpc, "return _G.got").await, Some(7));
}

#[tokio::test]
async fn reactions_are_async_never_inline() {
    let (rpc, _incoming) = start().await;
    // Even an already-resolved promise runs its :next on a later microtask, so
    // 'direct' (after the :next call) is recorded before 'reaction'.
    exec_lua(
        &rpc,
        "_G.order = {}\n\
         btv.promise.resolve(1):next(function() table.insert(_G.order, 'reaction') end)\n\
         table.insert(_G.order, 'direct')",
    )
    .await;
    assert_eq!(
        lua_string(&rpc, "return table.concat(_G.order, ',')")
            .await
            .as_deref(),
        Some("direct,reaction")
    );
}

// ----- combinators -----------------------------------------------------------

#[tokio::test]
async fn all_collects_values_in_order() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.sum = nil\n\
         btv.promise.all({ btv.promise.resolve(1), btv.promise.resolve(2), 3 })\n\
           :next(function(vs) _G.sum = vs[1] * 100 + vs[2] * 10 + vs[3] end)",
    )
    .await;
    // Plain (non-promise) values pass straight through, in input order.
    assert_eq!(lua_u64(&rpc, "return _G.sum").await, Some(123));
}

#[tokio::test]
async fn all_rejects_on_first_rejection() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.err = nil\n\
         btv.promise.all({ btv.promise.resolve(1), btv.promise.reject('nope') })\n\
           :catch(function(e) _G.err = e end)",
    )
    .await;
    assert_eq!(
        lua_string(&rpc, "return _G.err").await.as_deref(),
        Some("nope")
    );
}

#[tokio::test]
async fn all_settled_reports_each_outcome() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.report = nil\n\
         btv.promise.all_settled({ btv.promise.resolve(9), btv.promise.reject('x') })\n\
           :next(function(rs)\n\
             _G.report = rs[1].status .. ':' .. tostring(rs[1].value)\n\
               .. '|' .. rs[2].status .. ':' .. tostring(rs[2].reason)\n\
           end)",
    )
    .await;
    assert_eq!(
        lua_string(&rpc, "return _G.report").await.as_deref(),
        Some("fulfilled:9|rejected:x")
    );
}

#[tokio::test]
async fn any_takes_the_first_fulfilment() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.got = nil\n\
         btv.promise.any({ btv.promise.reject('a'), btv.promise.resolve(5) })\n\
           :next(function(v) _G.got = v end)",
    )
    .await;
    assert_eq!(lua_u64(&rpc, "return _G.got").await, Some(5));
}

// ----- btv.async / btv.await ---------------------------------------------------

#[tokio::test]
async fn async_await_reads_top_to_bottom() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.got = nil\n\
         local run = btv.async(function(start)\n\
           local a = btv.await(btv.promise.resolve(start))\n\
           local b = btv.await(btv.promise.resolve(a + 10))\n\
           return a + b\n\
         end)\n\
         run(1):next(function(v) _G.got = v end)",
    )
    .await;
    // a = 1, b = 11, result 12.
    assert_eq!(lua_u64(&rpc, "return _G.got").await, Some(12));
}

#[tokio::test]
async fn await_of_a_rejection_rejects_the_async_result() {
    let (rpc, _incoming) = start().await;
    // A rejected await re-raises inside the coroutine; uncaught, it rejects the
    // promise btv.async returned — caught by :catch on the result.
    exec_lua(
        &rpc,
        "_G.err = nil\n\
         local run = btv.async(function()\n\
           btv.await(btv.promise.reject('deep'))\n\
           return 'unreached'\n\
         end)\n\
         run():catch(function(e) _G.err = e end)",
    )
    .await;
    assert_eq!(
        lua_bool(
            &rpc,
            "return _G.err ~= nil and string.find(_G.err, 'deep') ~= nil"
        )
        .await,
        Some(true)
    );
}

#[tokio::test]
async fn await_outside_async_fails_loud() {
    let (rpc, _incoming) = start().await;
    // Per the no-silent-stubs rule: awaiting off a coroutine raises a named error.
    let ok = lua_bool(
        &rpc,
        "local ok, err = pcall(function() return btv.await(btv.promise.resolve(1)) end)\n\
         return (not ok) and string.find(err, 'btv.await') ~= nil",
    )
    .await;
    assert_eq!(ok, Some(true));
}

// ----- off-tick: btv.promise.delay --------------------------------------------

#[tokio::test]
async fn delay_fulfils_after_the_wall_clock_wait() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.done = false\n\
         btv.promise.delay(40, 'v'):next(function(v) _G.done = (v == 'v') end)",
    )
    .await;
    // Barrier #1: not yet — delay rides the loop, off the input tick.
    assert_eq!(lua_bool(&rpc, "return _G.done").await, Some(false));
    tokio::time::sleep(Duration::from_millis(150)).await;
    assert_eq!(lua_bool(&rpc, "return _G.done").await, Some(true));
}

// ----- registry hygiene ------------------------------------------------------

#[tokio::test]
async fn settled_promise_chains_do_not_leak_callbacks() {
    let (rpc, _incoming) = start().await;
    // A burst of fully-settled chains must leave the shared callback registry
    // empty — every microtask is a one-shot dropped after it fires.
    exec_lua(
        &rpc,
        "for _ = 1, 32 do\n\
           btv.promise.resolve(1):next(function(v) return v end):catch(function() end)\n\
         end",
    )
    .await;
    assert_eq!(
        lua_u64(
            &rpc,
            "local n = 0; for _ in pairs(btv._cb_fns) do n = n + 1 end; return n"
        )
        .await,
        Some(0)
    );
}

// ----- btv.promise.try (fold a sync throw + async result into one chain) -------

#[tokio::test]
async fn try_runs_fn_and_resolves_with_its_value() {
    let (rpc, _incoming) = start().await;
    // A plain return value flows through like btv.promise.resolve(value).
    exec_lua(
        &rpc,
        "_G.got = nil\n\
         btv.promise.try(function(a, b) return a + b end, 2, 3):next(function(v) _G.got = v end)",
    )
    .await;
    assert_eq!(lua_u64(&rpc, "return _G.got").await, Some(5));
}

#[tokio::test]
async fn try_turns_a_synchronous_throw_into_a_rejection() {
    let (rpc, _incoming) = start().await;
    // A function that errors SYNCHRONOUSLY (before returning) rejects the promise —
    // caught by the same :catch that an async rejection would land in.
    exec_lua(
        &rpc,
        "_G.err = nil\n\
         btv.promise.try(function() error('boom') end):catch(function(e) _G.err = tostring(e) end)",
    )
    .await;
    assert_eq!(
        lua_bool(&rpc, "return _G.err ~= nil and _G.err:find('boom') ~= nil").await,
        Some(true),
        "a sync throw is caught as a rejection"
    );
}

#[tokio::test]
async fn try_adopts_a_returned_promise() {
    let (rpc, _incoming) = start().await;
    // When fn returns a promise, try adopts it (waits on it) rather than fulfilling
    // with the promise object — so one chain handles sync-throw and async alike.
    exec_lua(
        &rpc,
        "_G.got = nil\n\
         btv.promise.try(function() return btv.promise.resolve(99) end):next(function(v) _G.got = v end)",
    )
    .await;
    assert_eq!(lua_u64(&rpc, "return _G.got").await, Some(99));
}

// ----- btv.on_next_tick / btv.wait_for: cross-tick deferral --------------------

// btv.on_next_tick defers to a LATER tick — not inline, not the same convergence.
#[tokio::test]
async fn on_next_tick_defers_to_a_later_tick() {
    let (rpc, _incoming) = start().await;
    // Arming it does NOT run it inline: `inline` captured right after is still false.
    exec_lua(
        &rpc,
        "_G.ran = false\n\
         btv.on_next_tick(function() _G.ran = true end)\n\
         _G.inline = _G.ran",
    )
    .await;
    assert_eq!(
        lua_bool(&rpc, "return _G.inline").await,
        Some(false),
        "on_next_tick must not run its fn inline"
    );
    // It fires on a later loop turn (within a short wall-clock window).
    tokio::time::sleep(Duration::from_millis(100)).await;
    assert_eq!(
        lua_bool(&rpc, "return _G.ran").await,
        Some(true),
        "on_next_tick fn should have run on a later tick"
    );
}

// btv.wait_for polls a predicate across ticks and fulfils with its truthy value.
#[tokio::test]
async fn wait_for_polls_across_ticks_and_resolves_with_the_value() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.n = 0\n\
         _G.got = nil\n\
         btv.wait_for(function()\n\
           _G.n = _G.n + 1\n\
           if _G.n >= 3 then return 'ready' end\n\
         end):next(function(v) _G.got = v end)",
    )
    .await;
    tokio::time::sleep(Duration::from_millis(200)).await;
    assert_eq!(
        lua_string(&rpc, "return _G.got").await,
        Some("ready".to_string()),
        "wait_for resolves with the predicate's truthy value once it holds"
    );
    assert_eq!(
        lua_bool(&rpc, "return _G.n >= 3").await,
        Some(true),
        "the predicate is polled across several ticks"
    );
}

// btv.wait_for rejects (with the given message) when the condition never holds within
// the bounded `tries`.
#[tokio::test]
async fn wait_for_rejects_after_bounded_tries() {
    let (rpc, _incoming) = start().await;
    exec_lua(
        &rpc,
        "_G.err = nil\n\
         btv.wait_for(function() return false end, { tries = 3, message = 'nope' })\n\
           :catch(function(e) _G.err = tostring(e and e.message or e) end)",
    )
    .await;
    tokio::time::sleep(Duration::from_millis(200)).await;
    assert_eq!(
        lua_string(&rpc, "return _G.err").await,
        Some("nope".to_string()),
        "an exhausted wait_for rejects with the configured message"
    );
}
